src/scratchOrgSettingsGenerator.ts

```diff
@@ -0,0 +1,421 @@
+import * as path from 'path';
+import { promises as fs } from 'fs';
+
+// @salesforce
+import { set, isEmpty, env, upperFirst } from '@salesforce/kit';
+import { ComponentSet, ComponentStatus, FileResponse } from '@salesforce/source-deploy-retrieve';
+import { get, getObject, JsonMap, ensureString, ensureObject, Nullable, Optional } from '@salesforce/ts-types';
+
+// Thirdparty
+import * as js2xmlparser from 'js2xmlparser';
+
+// Local
+import { Logger } from './logger';
+import { Messages } from './messages';
+import { SfdxError } from './sfdxError';
+import getApiVersion from './config/getApiVersion';
+import { writeJSONasXML } from './util/jsonXmlTools';
+import OrgPrefRegistry = require('./orgPrefRegistry');
+import { ScratchOrgInfo } from './scratchOrgInfoApi';
+
+Messages.importMessagesDirectory(__dirname);
+const messages: Messages = Messages.loadMessages('@salesforce/core', 'scratchOrgSettingsGenerator');
+
+interface ObjectSetting extends JsonMap {
+  sharingModel?: string;
+  defaultRecordType?: string;
+}
+
+export interface ScratchDefinition extends ScratchOrgInfo {
+  settings: Optional<Record<string, unknown>>;
+  objectSettings: Optional<Record<string, unknown>>;
+  orgPreferences: Optional<{
+    enabled: string[];
+    disabled: string[];
+  }>;
+}
+
+interface ObjectToBusinessProcessPicklist {
+  [key: string]: {
+    fullName: string;
+    default?: boolean;
+  };
+}
+
+interface BusinessProcessFileContent extends JsonMap {
+  fullName: string;
+  isActive: boolean;
+  values: [
+    {
+      fullName: string;
+      default?: boolean;
+    }
+  ];
+}
+
+/**
+ * Helper class for dealing with the settings that are defined in a scratch definition file.  This class knows how to extract the
+ * settings from the definition, how to expand them into a MD directory and how to generate a package.xml.
+ */
+export default class SettingsGenerator {
+  private settingData: Nullable<Record<string, unknown>>;
+  private objectSettingsData: Nullable<{ [objectName: string]: ObjectSetting }>;
+  private currentApiVersion = ensureString(getApiVersion());
+  private logger: Logger;
+
+  public constructor() {
+    this.logger = Logger.childFromRoot('SettingsGenerator');
+  }
+
+  /** extract the settings from the scratch def file, if they are present. */
+  public async extract(scratchDef: ScratchDefinition, apiVersion?: string): Promise<void> {
+    this.logger.debug('extracting settings from scratch definition file');
+    if (!apiVersion) {
+      apiVersion = this.currentApiVersion;
+    }
+    if (parseFloat(apiVersion) >= 47.0 && this.orgPreferenceSettingsMigrationRequired(scratchDef)) {
+      this.extractAndMigrateSettings(scratchDef);
+    } else {
+      this.settingData = getObject(scratchDef, 'settings');
+      this.objectSettingsData = getObject(scratchDef, 'objectSettings');
+    }
+
+    this.logger.debug('settings are', this.settingData);
+    // TODO, this is where we will validate the settings.
+    // See W-5068155
```

Comment:
@mshanemc The work item referenced here has been 'nevered'. Is that something you agree with? If so, then these comments can be removed.